### ISSUE TITLE
class 'cobbler.cexceptions.CX

### ISSUE THREAD

**Issue by QiChangYin** [cobbler/cobbler#3529]
接口1self.__sync_cobbler() 连续提交5次以上，就报上面的错误。
xmlrpc.client.Fault: <Fault 1: "<class 'cobbler.cexceptions.CX'>:'Cobbler trigger failed: cobbler.modules.sync_post_restart_services'">

而且把DHCP.service 带挂了，体验非常不好，能否优化一下提示sync的并发度。
作为用户根本不关心当前时间点有多少人执行sync，只关心本次提交是否ok。

**Comment by QiChangYin** [comment]
你们可以使用redis或者消息队列提示并发度。
1、当秒发流量低于5000以下时，可以使用消息队列异步处理，进行顺序消费，并且在消费，sync时增加锁机制，dhcp按照顺序单词执行sync即可。
2、当秒发流量非常大的时候，使用Redis记录入口时间，以秒为单位，启动异步任务轮训检查redis是否存在上一秒或者前几秒的记录，如果没有则跳过，如果有则消费，无论上一秒/前几秒进来多少sync请求，cobbler server只执行一次sync即可。

**Comment by SchoolGuy** [comment]
@QiChangYin English please... Otherwise I need to close the bug report.

**Comment by QiChangYin** [comment]
When the interface self.__sync_cobbler() is consecutively submitted more than 10 times, the following error is reported:

xmlrpc.client.Fault: <Fault 1: "<class 'cobbler.cexceptions.CX'>:'Cobbler trigger failed: cobbler.modules.sync_post_restart_services'">

And it caused disruption to the dhcpd.service, resulting in a poor user experience. Could you optimize it to increase the concurrency of the self.__sync_cobbler() execution? Users are not concerned about how many people are currently performing sync at a given moment; they are only interested in whether the current submission is successful and whether the XMLRPC can return normally.

**Comment by QiChangYin** [comment]
You can use Redis caching or a message queue to enhance the concurrency of the self.__sync_cobbler() operation.

When the traffic is below 5000 requests per second, consider using a message queue for asynchronous sequential consumption. Introduce a lock mechanism during the execution of sync, ensuring that dhcpd.service performs a single sync operation in sequence.

In high-traffic scenarios, utilize Redis to record entry times in seconds. Implement an asynchronous task to periodically check Redis for records from the previous second or seconds. If found, consume the records. Regardless of the number of sync requests entering in the previous second or seconds, execute only one sync operation for both Cobbler server and dhcpd.service.

**Comment by SchoolGuy** [comment]
Thank you for describing the problem in English again! That helps a lot for me. I don't trust Google Translate enough to correctly translate such a complex language for domain-specific problems.

Conceptionally a `cobbler sync` (user perspective) is a destructive operation that cannot run in parallel as it is rebuilding all files Cobbler touches from scratch. Furthermore re-executing this multiple hundred times seems like a user error to me.

There is work underway to substantially increase the performance of Cobbler. Sadly the number of developers that submit major feature work is not proportional to the amount of work to implement an asynchronous message queue in the near future. A queue would be a major design change that must be reflected in all clients which complicates things further. Lastly, a queue brings the need to use a document merging algorithm which not even Google get's right in all cases. So while I fully understand your request and would love to implement it, I will use this only as a very very last resort.

**Comment by QiChangYin** [comment]
Thank you very much for your answer. Our current solution to this problem is to solve it from the client and control the number of requests from the client.
You can use Redis instead of message queue, similar to the following solution

In high-traffic scenarios, utilize Redis to record entry times in seconds. Implement an asynchronous task to periodically check Redis for records from the previous second or seconds. If found, consume the records. Regardless of the number of sync requests entering in the previous second or seconds, execute only one sync operation for both Cobbler server and dhcpd.service.

**Comment by QiChangYin** [comment]
Hello, I have another issue to report. When deploying machines in batches of approximately 50 to 200 units, our platform concurrently executes and merges them into a single sync submission to Cobbler. We observed that Cobbler returns success, but subsequently, it disrupts DHCP. Additionally, when DHCP goes down, we noticed that it doesn't automatically recover.

**Comment by SchoolGuy** [comment]
@QiChangYin Cobbler automatically restarts DHCP. As such a short disruption (under 10 seconds) is to be expected. If the restart is not successful then the `cobbler sync` is marked as "failed". Furthermore, in the logfile of DHCP it should be visible what the reason for the failure is. Most of the time the reason is an error in the template.

**Comment by QiChangYin** [comment]
@SchoolGuy 

Thank you for your response.

In fact, the data within the Cobbler system is correct, so it's unexpected to encounter errors in the DHCP template. Theoretically, there should be no errors. The issue seems to resolve itself on the second attempt during sync.

For such scenarios, could we implement a pre-check mechanism, and add compensation or fallback mechanisms? If template errors are detected, the system could trigger another sync at the system level, ensuring that the correct template information is generated eventually.
